python/hail/linalg/blockmatrix.py

```diff
@@ -0,0 +1,483 @@
+
+        Parameters
+        ----------
+
+        cols_to_keep: :obj:`list` of :obj:`int`
+            Indices of columns to keep. Must be non-empty and increasing.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.filterCols(jarray(Env.jvm().long, cols_to_keep)))
+
+    @typecheck_method(rows_to_keep=listof(int),
+                      cols_to_keep=listof(int))
+    def filter(self, rows_to_keep, cols_to_keep):
+        """Filter matrix rows and columns.
+
+        Notes
+        -----
+
+        This method has the same effect as :meth:`BlockMatrix.filter_cols`
+        followed by :meth:`BlockMatrix.filter_rows` (or vice versa), but
+        filters the block matrix in a single pass which may be more efficient.
+
+        Parameters
+        ----------
+        rows_to_keep: :obj:`list` of :obj:`int`
+            Indices of rows to keep. Must be non-empty and increasing.
+        cols_to_keep: :obj:`list` of :obj:`int`
+            Indices of columns to keep. Must be non-empty and increasing.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.filter(jarray(Env.jvm().long, rows_to_keep),
+                                            jarray(Env.jvm().long, cols_to_keep)))
+
+
+    def to_numpy(self):
+        """Create a `NumPy ndarray
+        <https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html>`__
+        by collecting the block matrix.
+
+        Warning
+        -------
+
+        This method is not efficient for large matrices.
+
+        Returns
+        -------
+        NumPy ``ndarray``
+            ``ndarray`` with shape ``(n_rows, n_cols)``.
+        """
+
+        return numpy_from_breeze(self._jbm.toBreezeMatrix())
+
+    @property
+    def T(self):
+        """Matrix transpose.
+
+        Returns
+        -------
+        :class:`.LocalMatrix`
+        """
+
+        return BlockMatrix(self._jbm.transpose())
+
+    def cache(self):
+        """Cache the RDD of blocks.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.cache())
+
+    @typecheck_method(storage_level=storage_level)
+    def persist(self, storage_level='MEMORY_AND_DISK'):
+        """Persist the RDD of blocks.
+
+        Parameters
+        ----------
+        storage_level: :str:
+          Storage level for persistence.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.persist(storage_level))
+
+    def unpersist(self):
+        """Unpersist the RDD of blocks.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.unpersist())
+
+    @typecheck_method(b=block_matrix_type)
+    def __add__(self, b):
+        """Addition: A + B.
+
+        Parameters
+        ----------
+
+        b: :class:`BlockMatrix`
```

Comment:
no scalar addition/subtraction?

Reply:
Not yet. Here I'm documenting the functions that were already present. I plan to add all manner of broadcasted +,-,*,/ in subsequent PR

Reply:
Okay, I threw it in. I'll handle the reverse ops subsequently.